src/gfnff/gfnff_setup.f90

```diff
@@ -218,4 +219,28 @@ subroutine gfnff_input(env, mol, topo)
 
 end subroutine gfnff_input
 
+
+subroutine write_gfnff_adjacency(fname, topo)
+   implicit none
+   character(len=*),intent(in) :: fname
+   integer :: ifile ! file handle
+   type(TGFFTopology) :: topo
+   integer :: i, j
+
+   call open_file(ifile,fname,'w') 
+   ! looping over topology neighboring list
+   if (ifile.ne.-1) then
+      write(ifile, *) '# indices of neighbouring atoms (max seven); connectivity given by last entry;'  
+      do i = 1, size(topo%nb,2) ! #atoms
+         do j = 1, size(topo%nb,1) ! 20
+            write(ifile, '(I2,X)', advance='no') topo%nb(j,i)
```

Comment:
This statement has several issues, first since you are using a non-advancing output it will produce a trailing whitespace from the format specifier. Second, leaving away the width of the `x` identifier is a standard extension and we want to avoid relying on those.

Third, the integer field will overflow for more than 99 atoms:

```
❯ tail gfnff_adjacency
** ** **  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  3  
** ** **  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  3  
** ** **  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  3  
** ** **  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  3  
**  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  1  
**  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  1  
**  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  1  
**  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  1  
**  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  1  
86  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  1
```

Finally, printing the padded adjacency list exposes an implementation detail like the preallocated number of entries for each atom, which is fragile in case we change this internally. Maybe a more practical format might be

```f90
do i = 1, size(topo%nb, 2)
  write(ifile, '(*(i0:, 1x))') topo%nb(20, i), (topo%nb(j, i), j = 1, topo%nb(20, i))
end do
```

Reply:
Good idea, I adapted the output format.